src/main/java/com/amazon/dlic/auth/http/jwt/HTTPJwtAuthenticator.java

```diff
@@ -228,19 +230,27 @@ protected String extractSubject(final Claims claims, final RestRequest request)
     }
 
     @SuppressWarnings("unchecked")
-    protected String[] extractRoles(final Claims claims, final RestRequest request) {
-    	// no roles key specified
+    protected String[] extractRoles(final Claims claims) {
+        String[] roles = new String[0];
+        // no roles key specified
     	if(rolesKey == null) {
-    		return new String[0];
-    	}
-		// try to get roles from claims, first as Object to avoid having to catch the ExpectedTypeException
-    	final Object rolesObject = claims.get(rolesKey, Object.class);
-    	if(rolesObject == null) {
-    		log.warn("Failed to get roles from JWT claims with roles_key '{}'. Check if this key is correct and available in the JWT payload.", rolesKey);
-    		return new String[0];
+    		return roles;
     	}
 
-    	String[] roles = String.valueOf(rolesObject).split(",");
+    	Object rolesObject;
+    	if (rolesKey.contains(".")){
```

Comment:
Question for my understanding: 
Can someone have "." in their roleskey  but not having actual nested roles?  
 Is there any other IdP who allows "." in their rolekey name without actually having nested roles?

Reply:
I am not sure about other IdPs (we only use keycloak).
The keycloak default uses the "." for nested if you provide a custom claim. The default one is always nested.
![image](https://user-images.githubusercontent.com/13332383/106102576-e489c480-613f-11eb-8282-6dcd433790d1.png)

Reply:
The main reason for this PR is that all examples in the [documentation ](https://opendistro.github.io/for-elasticsearch-docs/docs/security/configuration/openid-connect/#configure-openid-connect-integration) use keycloak.
But the examples will not work without custom claims that map the roles into a new claim.

Reply:
Based on your screenshot information, I think this IdP also supports using '.' literally into key name by escaping. That means key names can have '.' in their name but it will not be nested if '.' was escaped using '\'. 

In this case to avoid breaking customers who are using '.' literally in their key name, I think we should introduce some flag in configuration for nested claims. (like is_roles_key_nested: true  | false with default set to false). 

For ex.,

authc:
    openid_auth_domain:
        http_enabled: true
        transport_enabled: true
        order: 0
        http_authenticator:
        type: openid
        challenge: false
        config:
            subject_key: preferred_username
            roles_key: realm_access.roles
            is_roles_key_nested: true 
            openid_connect_url: https://my-keycloak-server.com:8443/auth/realms/demo/.well-known/openid-configuration
        authentication_backend:
        type: noop


thoughts?

Reply:
Nested Roles can be used like this in the configuration 

A JSON path expression to the payload that stores the user’s roles, for example $["foo"]["user"]["roles"] where foo is the claim name in the token 

In your case, did you try adding secuirty configuration as ["realm_access"]["roles"]?

Reply:
I think roles_path is a good idea. However, I am not sure if I will be able to implement this and create the corresponding tests in the near future (due to limited time).

Reply:
I can take this up and add it to my tasks.

Reply:
Created issue from 

> Just realised we dont have this functionality
> 
> We should introduced roles_path
> 
> `
> 
> ```java
> if (jsonRolesPath != null) {
>                 Object roles = JsonPath.using(jsonPathConfig).parse(claims.asMap()).read(jsonRolesPath);
>                 return splitRoles(roles);
>         
> ```
> 
> This way customer using the "." in their key in payload will not be affected

Created issue using this to track separately: https://github.com/opendistro-for-elasticsearch/security/issues/1003  

We can close this PR, if we are agree to this.  Thanks all.